File: backend/app/schemas/report.py
"""舆情报告响应模型（P2 报告自动生成 + PDF 导出 / Phase Report-1 可配置生成器）。"""
from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional, Union

from pydantic import BaseModel, Field

from app.schemas.dashboard import (
    KeywordItem,
    RegionItem,
    SentimentItem,
    SourceItem,
    TrendItem,
)


class ReportOpinionItem(BaseModel):
    """高风险舆情 TOP 条目。"""

    id: int
    title: str
    source: str
    region_name: str
    risk_score: int
    sentiment: str
    created_at: str
    summary: str


class ReportEventItem(BaseModel):
    """重点事件条目。"""

    id: int
    title: str
    risk_level: str
    opinion_count: int


class ReportOverviewResponse(BaseModel):
    """舆情报告总览（JSON，供前端预览 / PDF 渲染共用）。"""

    generated_at: str
    period_days: int
    total: int
    today: int
    high_risk: int
    event_count: int
    risk_rate: float
    negative_rate: float
    trend: List[TrendItem]
    top_keywords: List[KeywordItem]
    top_sources: List[SourceItem]
    top_regions: List[RegionItem]
    top_risky: List[ReportOpinionItem]
    events: List[ReportEventItem]
    sentiments: List[SentimentItem]


# ---------------------------------------------------------------------------
# Phase Report-1：可配置报告生成器
# ---------------------------------------------------------------------------
class ReportModuleParamDef(BaseModel):
    """模块可配置参数的元信息（供前端渲染表单）。"""

    key: str
    label: str
    type: str = "int"
    default: Any = None
    min: Optional[int] = None
    max: Optional[int] = None


class ReportModuleDef(BaseModel):
    """单个可选报告模块的元信息。

    Phase Report-2-P1 新增 name / default_enabled / params；title 保留不变，
    Phase Report-1.1 前端（读 key/title/description）零破坏。
    """

    key: str
    name: str
    title: str
    description: str
    default_enabled: bool = True
    params: List[ReportModuleParamDef] = Field(default_factory=list)


class ReportModulesResponse(BaseModel):
    """GET /reports/modules 返回：所有可选模块 + 默认选中的模块。"""

    modules: List[ReportModuleDef]
    default_modules: List[str]


class ReportGenerateRequest(BaseModel):
    """POST /reports/generate 请求体（Legacy，Phase Report-1.1 已上线）。"""

    report_name: str = "舆情监测报告"
    time_field: Literal["created_at", "publish_time"] = "created_at"
    start_date: Optional[str] = None  # "YYYY-MM-DD"
    end_date: Optional[str] = None
    days: int = 7
    module_keys: List[str]


# ---------------------------------------------------------------------------
# Phase Report-2-P1：正式导出入口 POST /reports/export
# ---------------------------------------------------------------------------
class ReportModuleSelection(BaseModel):
    """带参数的模块选择项（modules 支持 "key" 字符串或本对象两种写法）。"""

    key: str
    params: Dict[str, Any] = Field(default_factory=dict)


class ReportExportRequest(BaseModel):
    """POST /reports/export 请求体。

    时间口径为本地（Asia/Shanghai）日期语义，不做任何时区转换
    （Phase Report-2-P0 决策：方案 A）。
    """

    name: str = "舆情监测报告"
    time_field: Literal["created_at", "publish_time"] = "created_at"
    range_type: Literal["last_n_days", "custom"] = "last_n_days"
    range_days: int = Field(default=7, ge=1, le=365)
    start_date: Optional[str] = None  # "YYYY-MM-DD"（range_type=custom 必填）
    end_date: Optional[str] = None
    modules: List[Union[str, ReportModuleSelection]] = Field(default_factory=list)
    # 当前阶段仅支持 download；email 留作 Phase 5 邮件能力，此处显式拒绝。
    delivery: Literal["download", "email"] = "download"
    recipients: List[str] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Phase Report-4-A：报告模板（最小可用：保存/加载）
# ---------------------------------------------------------------------------
class ReportTemplateConfig(BaseModel):
    """模板保存的「导出配置快照」结构。

    与 ReportExportRequest 同构，但剔除 delivery / recipients
    （模板只描述「怎么生成」，不描述「怎么投递」）。
    """

    name: str = "舆情监测报告"
    time_field: Literal["created_at", "publish_time"] = "created_at"
    range_type: Literal["last_n_days", "custom"] = "last_n_days"
    range_days: int = Field(default=7, ge=1, le=365)
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    modules: List[Union[str, ReportModuleSelection]] = Field(default_factory=list)


class ReportTemplateCreate(BaseModel):
    """POST /reports/templates 请求体。"""

    name: str = Field(..., min_length=1, max_length=128)
    description: Optional[str] = Field(default=None, max_length=255)
    is_public: bool = False
    config_json: ReportTemplateConfig


class ReportTemplateUpdate(BaseModel):
    """PUT /reports/templates/{id} 请求体（全字段可选）。"""

    name: Optional[str] = Field(default=None, min_length=1, max_length=128)
    description: Optional[str] = Field(default=None, max_length=255)
    is_public: Optional[bool] = None
    config_json: Optional[ReportTemplateConfig] = None


class ReportTemplateResponse(BaseModel):
    """模板对外返回（含 can_edit 标记，供前端决定是否展示删除/编辑）。"""

    id: int
    name: str
    description: Optional[str] = None
    owner_id: int
    config_json: ReportTemplateConfig
    is_public: bool
    created_at: str
    updated_at: str
    can_edit: bool

File: frontend/src/api/report.ts
import api from './index'

/** 模块可配置参数的元信息（由后端 GET /reports/modules 下发，供前端渲染表单）。 */
export interface ReportModuleParamDef {
  key: string
  label: string
  type: string
  default: any
  min?: number | null
  max?: number | null
}

export interface ReportModuleDef {
  key: string
  name: string
  title: string
  description: string
  default_enabled: boolean
  params: ReportModuleParamDef[]
}

export interface ReportModulesResp {
  modules: ReportModuleDef[]
  default_modules: string[]
}

/** 模块选择项：可直接传 key 字符串，或带参数的对象（后端 Union[str, {key, params}]）。 */
export type ReportModuleSelection = string | { key: string; params: Record<string, any> }

/** POST /reports/export 请求体（与后端 ReportExportRequest 对齐）。 */
export interface ReportExportPayload {
  name: string
  time_field: 'created_at' | 'publish_time'
  range_type: 'last_n_days' | 'custom'
  range_days: number
  start_date: string | null
  end_date: string | null
  modules: ReportModuleSelection[]
  delivery: 'download'
}

/** 获取可配置报告的可选模块清单（含默认选中项与参数元信息）。 */
export function getReportModules() {
  return api.get<ReportModulesResp>('/reports/modules')
}

/** 按自定义配置生成并下载 PDF 报告（返回 blob）。Phase Report-2-P1 正式入口。 */
export function generateReport(payload: ReportExportPayload) {
  return api.post('/reports/export', payload, { responseType: 'blob' })
}

// ===== Phase Report-4-A：报告模板（保存/加载） =====

/** 模板里的导出配置快照（与 ReportExportPayload 同构，不含 delivery）。 */
export interface ReportTemplateConfig {
  name: string
  time_field: 'created_at' | 'publish_time'
  range_type: 'last_n_days' | 'custom'
  range_days: number
  start_date: string | null
  end_date: string | null
  modules: ReportModuleSelection[]
}

/** 创建模板请求体。 */
export interface ReportTemplateCreatePayload {
  name: string
  description?: string | null
  is_public: boolean
  config_json: ReportTemplateConfig
}

/** 更新模板请求体（全字段可选）。 */
export interface ReportTemplateUpdatePayload {
  name?: string
  description?: string | null
  is_public?: boolean
  config_json?: ReportTemplateConfig
}

/** 模板对外结构（含 can_edit 标记）。 */
export interface ReportTemplate {
  id: number
  name: string
  description?: string | null
  owner_id: number
  config_json: ReportTemplateConfig
  is_public: boolean
  created_at: string
  updated_at: string
  can_edit: boolean
}

/** 加载当前用户可访问的模板（自己的 + 公共）。需 reports:export。 */
export function getTemplates() {
  return api.get<ReportTemplate[]>('/reports/templates')
}

/** 保存为模板。需 reports:manage。 */
export function createTemplate(payload: ReportTemplateCreatePayload) {
  return api.post<ReportTemplate>('/reports/templates', payload)
}

/** 更新模板。需 reports:manage（仅 owner/admin 可操作）。 */
export function updateTemplate(id: number, payload: ReportTemplateUpdatePayload) {
  return api.put<ReportTemplate>(`/reports/templates/${id}`, payload)
}

/** 删除模板。需 reports:manage（仅 owner/admin 可操作）。 */
export function deleteTemplate(id: number) {
  return api.delete(`/reports/templates/${id}`)
}
